### PR TITLE
(fix): make pipeline pickleable

### LINKED ISSUE #66: `CodecPipelineImpl` Not pickle-able
```python
import pickle
import zarrs
import zarr
import numpy as np
from pathlib import Path

zarr.config.set({"codec_pipeline.path": "zarrs.ZarrsCodecPipeline"})

arr = zarr.create(
    (2, 2),
    store=zarr.storage.LocalStore(root="foo.zarr"),
    chunks=(1, 1),
    dtype=np.int16,
    fill_value=10,
    codecs=[zarr.codecs.BytesCodec(), zarr.codecs.BloscCodec()],
)

with Path.open('pickled_pipeline', 'wb') as f:
    pickle.dump(arr._async_array.codec_pipeline, f)
```

Error:

```
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
TypeError: cannot pickle 'builtins.CodecPipelineImpl' object
```

Commit: e43bff057ccb6af88394b45aaf71ea853285054f

### PR DESCRIPTION
**Pull request by ilan-gold**
Fixed #66

I noticed this while using dask with process workers

### CODE REVIEW

File: python/zarrs/pipeline.py [hunk 1]
```diff
@@ -32,10 +32,41 @@
...
+        ),
+        # TODO: upstream zarr-python array.write_empty_chunks is not merged yet #2429
+        store_empty_chunks=config.get("array.write_empty_chunks", None),
+        chunk_concurrent_minimum=config.get(
+            "codec_pipeline.chunk_concurrent_minimum", None
+        ),
+        chunk_concurrent_maximum=config.get(
+            "codec_pipeline.chunk_concurrent_maximum", None
+        ),
+        num_threads=config.get("threading.max_workers", None),
```
Comment (ilan-gold) [inline python/zarrs/pipeline.py:49]: Not sure even what the promise is from `zarr` re: the stability of values fetched from the `config` so not sure if this is a potential source of bugs or not

File: python/zarrs/pipeline.py [hunk 2]
```diff
@@ -32,10 +32,41 @@
...
+    )
+
+class ZarrsCodecPipelineState(TypedDict):
+    codec_metadata_json: str
+    codecs: tuple[Codec, ...]
+
+@dataclass
```
Comment (ilan-gold) [inline python/zarrs/pipeline.py:56]: Any reason to have this frozen from a practical perspective?

File: python/zarrs/pipeline.py [hunk 3]
```diff
@@ -32,10 +32,40 @@
...
+        # TODO: upstream zarr-python array.write_empty_chunks is not merged yet #2429
+        store_empty_chunks=config.get("array.write_empty_chunks", None),
+        chunk_concurrent_minimum=config.get(
+            "codec_pipeline.chunk_concurrent_minimum", None
+        ),
+        chunk_concurrent_maximum=config.get(
+            "codec_pipeline.chunk_concurrent_maximum", None
+        ),
+        num_threads=config.get("threading.max_workers", None),
+    )
```
Comment (ilan-gold) [inline python/zarrs/pipeline.py:48]: Not sure even what the promise is from zarr re: the stability of values fetched from the config so not sure if this is a potential source of bugs or not

Comment (LDeakin) [reply python/zarrs/pipeline.py:48]: It is still up in the air: see https://github.com/zarr-developers/zarr-python/pull/2429#discussion_r1811364395

**Review by LDeakin — APPROVED** [review]

File: python/zarrs/pipeline.py (hunk 3 above)
Comment (ilan-gold) [reply python/zarrs/pipeline.py:48]: I'll add a comment with a warning for the next person then in our docs about this fact and then maybe ask on the zulip but I think that shouldn't block this PR

Comment (ilan-gold) [reply python/zarrs/pipeline.py:48]: See: https://ossci.zulipchat.com/#narrow/channel/423692-Zarr/topic/Config.20Values.  Seems like it's best to leave this as-is for now

Comment (ilan-gold) [reply python/zarrs/pipeline.py:48]: Ok a good point raised about empty chunks: https://ossci.zulipchat.com/#narrow/channel/423692-Zarr/topic/Config.20Values/near/487023639

Probably should be pickled

Comment (ilan-gold) [reply python/zarrs/pipeline.py:48]: A good point raised: Given #**Zarr>Config Values@487049771**  

So will leave this as-is

**Pull request merged**